crates/uv-auth/src/middleware.rs

```diff
@@ -40,94 +40,55 @@ impl AuthMiddleware {
         next: Next<'_>,
     ) -> reqwest_middleware::Result<Response> {
-        let url = req.url().clone();
+        let url = request.url().clone();
 
         // If the request already has an authorization header, we don't need to do anything.
         // This gives in-URL credentials precedence over the netrc file.
-        if req.headers().contains_key(reqwest::header::AUTHORIZATION) {
-            debug!("Request already has an authorization header: {url}");
-            return next.run(req, _extensions).await;
+        if request
+            .headers()
+            .contains_key(reqwest::header::AUTHORIZATION)
+        {
+            debug!("Credentials are already present for {url}");
+            return next.run(request, extensions).await;
         }
 
-        // Try auth strategies in order of precedence:
-        if let Some(stored_auth) = GLOBAL_AUTH_STORE.get(&url) {
-            // If we've already seen this URL, we can use the stored credentials
-            if let Some(auth) = stored_auth {
-                debug!("Adding authentication to already-seen URL: {url}");
-                req.headers_mut().insert(
-                    reqwest::header::AUTHORIZATION,
-                    basic_auth(auth.username(), auth.password()),
-                );
-            } else {
-                debug!("No credentials found for already-seen URL: {url}");
-            }
-        } else if let Some(auth) = self.nrc.as_ref().and_then(|nrc| {
-            // If we find a matching entry in the netrc file, we can use it
-            url.host_str()
-                .and_then(|host| nrc.hosts.get(host).or_else(|| nrc.hosts.get("default")))
-        }) {
-            let auth = Credential::from(auth.to_owned());
-            req.headers_mut().insert(
-                reqwest::header::AUTHORIZATION,
-                basic_auth(auth.username(), auth.password()),
-            );
-            GLOBAL_AUTH_STORE.set(&url, Some(auth));
-        } else if matches!(self.keyring_provider, KeyringProvider::Subprocess) {
-            // If we have keyring support enabled, we check there as well
-            match get_keyring_subprocess_auth(&url) {
-                Ok(Some(auth)) => {
-                    req.headers_mut().insert(
-                        reqwest::header::AUTHORIZATION,
-                        basic_auth(auth.username(), auth.password()),
-                    );
-                    GLOBAL_AUTH_STORE.set(&url, Some(auth));
-                }
-                Ok(None) => {
-                    debug!("No keyring credentials found for {url}");
-                }
-                Err(e) => {
-                    warn!("Failed to get keyring credentials for {url}: {e}");
+        if let Some(credentials) = GLOBAL_AUTH_STORE.get(&url) {
+            debug!("Adding stored credentials to {url}");
+            request = credentials.authenticated_request(request)
+        } else if GLOBAL_AUTH_STORE.should_attempt_fetch(&url) {
```

Comment:
I think one minor downside here is we now do two hash lookups per unauthenticated request. I wonder if returning an enum from `.get` would make this simpler? Like three cases: we have credentials, the credentials are unset, there are no credetials.

Reply:
Yeah I considered this — an enum does seem helpful for that. Are we really concerned about that being a performance bottleneck though? I guess the mutex makes it even slower?

Reply:
I don't know if the perf matters, probably not, but I do find myself thinking that if we're querying the data source twice for the same information, something is not quite right in the data model.

Reply:
I'll switch to an enum it makes sense to me

Reply:
I still think the double fetch here represents a deficiency in the data model but perhaps it's addressed in the next PR.